h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoost.java

```diff
@@ -268,17 +271,28 @@ final void buildModelImpl() {
       } else if (_parms._dmatrix_type == XGBoostModel.XGBoostParameters.DMatrixType.dense) {
         model._output._sparse = false;
       } else {
-        float fillRatio = 0;
-        int col = 0;
+        BigDecimal nonZeroCount = BigDecimal.ZERO;
+        int nonCategoricalCols = 0;
+        long numOneHotEncodedCols = 0;
         for (int i = 0; i < _train.numCols(); ++i) {
           if (_train.name(i).equals(_parms._response_column)) continue;
           if (_train.name(i).equals(_parms._weights_column)) continue;
           if (_train.name(i).equals(_parms._fold_column)) continue;
           if (_train.name(i).equals(_parms._offset_column)) continue;
-          fillRatio += _train.vec(i).nzCnt() / _train.numRows();
-          col++;
+          // This may overflow and will need to be fixed in production code
+          numOneHotEncodedCols += Math.max(0, _train.vec(i).cardinality());
+          if(_train.vec(i).isCategorical()){
+            nonZeroCount = nonZeroCount.add(BigDecimal.valueOf(_train.numRows()));
+          }else {
+            nonZeroCount = nonZeroCount.add(BigDecimal.valueOf(_train.vec(i).nzCnt()));
+          }
+          if(!_train.vec(i).isCategorical()) {
+            nonCategoricalCols++;
+          }
         }
-        fillRatio /= col;
+        BigDecimal denominator = BigDecimal.valueOf(numOneHotEncodedCols).add(BigDecimal.valueOf(nonCategoricalCols));
+        denominator = denominator.multiply(BigDecimal.valueOf(_train.numRows()), MathContext.DECIMAL128);
+        double fillRatio = nonZeroCount.divide(denominator, MathContext.DECIMAL128).doubleValue();
         Log.info("fill ratio: " + fillRatio);
         model._output._sparse = fillRatio < 0.5 || ((_train.numRows() * (long) _train.numCols()) > Integer.MAX_VALUE);
```

Comment:
Pavel:  Don't quite understand the logic here.  Okay, basically, we want to switch over to a sparse representation (lists denoting the row and column indices of an element) instead of using a double array.  Hence, we have this fillRatio thing to guide us (may want to change this ratio depending on the memory requirement of both representations).  However, what is the second OR condition for?